[Merged by Bors] - refactor: change MvQPF to use bundled inheritance

alexkeizer opened the pull request:
There is a clear hierarchy of the type classes that apply to multivariate type functions `(F : TypeVec n -> Type)`. Namely, every `MvQPF` is a `LawfulMvFunctor` and every `LawfulMvFunctor` is a `MvFunctor`.

This commit changes to use bundled inheritance (i.e., `extends`) for `MvQPF`, so that we don't have to repeat the `[MvFunctor F]` assumption every time we want `[MvQPF F]`

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser reviewed (commented):
From the paper

> With these definitions and notation, the definition of a multivariate QPF is almost exactly the same as the definition of a unary one:

where the univariate one does not extend `Functor`:

https://github.com/leanprover-community/mathlib4/blob/64da86b2282d68ff0478b2b3f210fcd92bc02ca9/Mathlib/Data/QPF/Univariate/Basic.lean#L51-L56

so your new version breaks the symmetry:

https://github.com/leanprover-community/mathlib4/blob/64da86b2282d68ff0478b2b3f210fcd92bc02ca9/Mathlib/Data/QPF/Multivariate/Basic.lean#L85-L90

I don't think we should break this symmetry; which is either a vote against this change, or a vote in favor of simultaneously refactoring the univariate one.

eric-wieser commented:
@digama0, can you comment on this?

alexkeizer commented:
> I don't think we should break this symmetry; which is either a vote against this change, or a vote in favor of simultaneously refactoring the univariate one.

Fair point. I do think the bundled version has significantly nicer ergonomics, so for me that would mean a vote in favor of a simultaneous refactor

digama0 commented on Mathlib/Data/QPF/Multivariate/Constructions/Quot.lean:
```diff
@@ -28,20 +28,16 @@ variable {F : TypeVec.{u} n → Type u}
-def quotientQPF (FG_abs_repr : ∀ {α} (x : G α), FG_abs (FG_repr x) = x)
-    (FG_abs_map : ∀ {α β} (f : α ⟹ β) (x : F α), FG_abs (f <$$> x) = f <$$> FG_abs x) : MvQPF G
-    where
+def quotientQPF (F) [q : MvQPF F]
+                {G : TypeVec.{u} n → Type u} [MvFunctor G]
+                {FG_abs : ∀ {α}, F α → G α}
+                {FG_repr : ∀ {α}, G α → F α}
+                (FG_abs_repr : ∀ {α} (x : G α), FG_abs (FG_repr x) = x)
+                (FG_abs_map : ∀ {α β} (f : α ⟹ β) (x : F α), FG_abs (f <$$> x) = f <$$> FG_abs x) :
```
```suggestion
def quotientQPF (F) [q : MvQPF F]
    {G : TypeVec.{u} n → Type u} [MvFunctor G]
    {FG_abs : ∀ {α}, F α → G α}
    {FG_repr : ∀ {α}, G α → F α}
    (FG_abs_repr : ∀ {α} (x : G α), FG_abs (FG_repr x) = x)
    (FG_abs_map : ∀ {α β} (f : α ⟹ β) (x : F α), FG_abs (f <$$> x) = f <$$> FG_abs x) :
```
mathlib doesn't do "visual indent"

digama0 commented on Mathlib/Data/QPF/Multivariate/Constructions/Comp.lean:
```diff
@@ -55,34 +54,46 @@ protected theorem mk_get (x : (Comp F G) α) : Comp.mk (Comp.get x) = x := rfl
-  abs_map := by intros; simp only [(· ∘ ·)]; rw [← abs_map]
-                simp only [comp.get_map, map_map, TypeVec.comp, abs_map, map_mk]
+section
+  variable [MvFunctor F] [∀ i, MvFunctor <| G i]
```
sections are not indented

digama0 commented:
I have no opinion on the main change, it seems reasonable and I guess it decreases verbosity? It should not be bundled with the function `F` itself, because this needs to be constructible using lambda or equivalent for the (eventual) automation to make use of it, but `MvFunctor` vs `MvQPF` doesn't make much difference.

alexkeizer commented:
I've now refactored the univariate QPFs as well, and fixed the formatting.

This change indeed reduces verbosity, but that is not my main motivation.
The reason I originally made the refactor (in alexkeizer/QpfTypes) mostly has to with inference problems of the form `[forall i, MvQpf (Gs i)]`, where `Gs : Fin n -> TypeVec m -> Type*`. To work around some issues, I introduces a surrogate typeclass of the form
```lean
class VecMvQpf {n : Nat} (v : Fin n -> TypeVec m -> Type*) where
      prop : ∀ i, Qpf (v i)
```
Together with instances that essentially do induction on `i`. 

Having to deal with unbundled typeclass assumptions was quite cumbersome in this scenario.

alexkeizer commented:
I've just merged master and resolved the conflicts. @digama0 since you said the changes seem reasonable, and nobody has since objected, would you feel comfortable merging it?
If not, what can I do to make this easier to merge?

YaelDillies reviewed (approved):
This looks reasonable to me.

maintainer merge

fpvandoorn commented:
bors merge
